### PR TITLE
🪢 feat: Reference Previous Tool Outputs in Subsequent Calls

### PR DESCRIPTION
**Pull request by danny-avila**
## Summary

I added an opt-in mechanism that lets any tool pipe a previously-executed tool's output into its own arguments via a `{{tool<idx>turn<turn>}}` placeholder, modeled after the turn-based anchors web search already uses. The feature is gated by a new `RunConfig.toolOutputReferences` field and is disabled by default; size limits default to match the existing tool-result truncation budget so nothing substituted can exceed what the model already saw.

- Added a run-scoped `ToolOutputReferenceRegistry` that stores successful tool outputs under stable `tool<idx>turn<turn>` keys, with per-output clipping and FIFO eviction when the aggregate exceeds `maxTotalSize`.
- Wired the registry into `ToolNode.runTool` and `ToolNode.dispatchToolEvents`: increments a per-run turn counter once per batch, resolves placeholders in string args before invoke, and annotates the returned `ToolMessage.content` for the LLM.
- Annotated content is a `_ref` key injected into parsed-JSON-object outputs (with pretty-print preservation and collision fallback) or a `[ref: …]\n` prefix for strings / arrays / primitives — the stored registry value stays un-annotated so piping into `jq`/`grep`/`awk` produces pristine input.
- Surfaced unresolved placeholders via a trailing `[unresolved refs: …]` line on the same tool message so the model can self-correct.
- Threaded `toolOutputReferences` from `RunConfig` through `Run` into both `StandardGraph` and `MultiAgentGraph`, which pass it into every compiled `ToolNode`; subagent-as-tool calls benefit automatically since they flow through the same paths.
- Introduced `HARD_MAX_TOTAL_TOOL_OUTPUT_SIZE = 800_000` and a `calculateMaxTotalToolOutputSize(perOutput)` helper (`2× per-output, clamped`) alongside the existing `HARD_MAX_TOOL_RESULT_CHARS`.
- Updated the bash tool description with a dedicated section explaining the reference format, prefix vs `_ref` placement, how to embed placeholders, and what happens when a key is unknown.

## Change Type

- [x] New feature (non-breaking change which adds functionality)

### Test Configuration

- Node + Jest via `npx jest`, typecheck via `npx tsc --noEmit`, lint via `npx eslint`.

I added 34 new tests (22 unit + 12 integration) across two new files and ran the full `src/tools/` and `src/graphs/` suites plus typecheck and lint.

- `src/tools/__tests__/toolOutputReferences.test.ts` covers registry set/get, per-output clipping, FIFO eviction across limit breaches, deep-walked resolution in strings/nested objects/arrays, deduplicated unresolved reporting, `_ref` injection for JSON objects (including pretty-print preservation and collision fallback), `[ref: …]` fallback for arrays/primitives/parse failures, and the regex pattern.
- `src/tools/__tests__/ToolNode.outputReferences.test.ts` covers disabled-by-default behavior, prefix vs `_ref` annotation in `ToolMessage.content`, un-annotated registry storage (so `{{tool0turn0}}` substitutes clean output downstream), per-batch turn counter increments, per-batch tool index assignment, unresolved-ref reporting, per-output clipping in the registry, aggregate FIFO eviction across batches, and the fact that errored tool calls are not registered.

To exercise manually: enable in `RunConfig` with `toolOutputReferences: { enabled: true }`, run a sequence of bash tool calls where the second command contains `{{tool0turn0}}`, and confirm the substituted stdout flows in and a `[ref: tool1turn0]` prefix (or `_ref` field) appears on the follow-up result.

## Checklist

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented in any complex areas of my code
- [x] I have made pertinent documentation changes
- [x] My changes do not introduce new warnings
- [x] I have written tests demonstrating that my changes are effective or that my feature works
- [x] Local unit tests pass with my changes

### CODE REVIEW

**Comment by danny-avila** [comment]
@codex review

**Comment by danny-avila** [comment]
Thanks @chatgpt-codex-connector — valid catch on the `PreToolUse` re-resolution. Also ran a deeper self-audit and addressed several additional findings before requesting another pass:

**Fixes in this push (e2b13a8, 57d498f, de9b782):**
- **Re-resolve after `PreToolUse` rewrites args** — the direct cause you flagged. If `updatedInput` introduces new `{{…}}` placeholders, they now get substituted and the unresolved-refs record is refreshed so the LLM warning reflects the actually-sent args.
- **Bash description is now conditional** — the reference guidance moved from the static `BashExecutionToolDescription` constant into a separate `BashToolOutputReferencesGuide` + a `buildBashExecutionToolDescription({ enableToolOutputReferences })` composer. With the feature disabled (default), the LLM is no longer told to emit `{{tool0turn0}}` syntax that would leak into the shell.
- **Reset registry/turn counter on runId change** — a host reusing the same compiled workflow across `processStream` calls no longer serves stale `tool<i>turn<n>` from a previous run. Detected via `config.configurable.run_id` drift inside `run()`.
- **Handle `ToolMessage` returns** — the LangChain `tool()` helper auto-wraps string returns in a ToolMessage, which was taking the early-return branch and skipping annotation/registration entirely. The `tool`-typed branch now applies the reference key (string content) or logs a warning + passes through (array content, documented as a known limitation).
- **Renamed public accessor to `_unsafeGetToolOutputRegistry`** with `@internal`. Keeps it available for test observation but signals to library consumers that external mutation is unsupported.
- **Registry internals**: dropped redundant `key` field (now plain `Map<string, string>`), moved the stateful `/g` matcher off the module-level export (exported `TOOL_OUTPUT_REF_PATTERN` is now non-global — no `lastIndex` footgun), added a `hasAnyPlaceholder` pre-scan so placeholder-free args skip the deep walk. Fixed "appended" → "prepended" JSDoc, and documented `maxTotalSize` as a soft cap (most-recent entry is never self-evicted).

**New tests** (18 total in `ToolNode.outputReferences.test.ts`, up from 12):
- Event-driven dispatch path: annotation, host request substitution, unresolved-ref reporting, PostToolUse replacing output is what ends up registered, and PreToolUse re-resolution.
- Direct path: runId-change resets registry + turn counter and surfaces unresolved placeholders from the new run.

Tradeoffs I didn't change:
- `_ref` stays as the injected key. The collision fallback to `[ref: …]` prefix keeps the feature lossless, and a more exotic key like `__toolRef` adds tokens without measurable benefit.
- JSON round-trip on annotation stays unconditional. A size-gated fast path would help only for very large JSON outputs (approaching the 400KB cap); since the registry already clips at that budget, the perf concern is bounded.

Ready for another pass.

**Comment by danny-avila** [comment]
@codex review

File: src/tools/toolOutputReferences.ts [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+  private evictUntilWithinLimit(): void {
+    if (this.totalSize <= this.maxTotalSize) {
+      return;
+    }
+    for (const key of this.entries.keys()) {
+      if (this.totalSize <= this.maxTotalSize) {
+        return;
+      }
+      if (this.entries.size <= 1) {
+        return;
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/tools/toolOutputReferences.ts]: Fixed in 8e0973c — `maxOutputSize` is now clamped to `min(maxOutputSize, maxTotalSize)` at construction, so a single stored entry can never exceed the aggregate bound. The eviction loop no longer retains the just-stored entry when it alone violates the cap either. Updated the config JSDoc to drop the soft-cap language (it's now a hard cap), and added a test for the `maxOutputSize: 1000, maxTotalSize: 10` case you flagged.

File: src/tools/toolOutputReferences.ts [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+    obj[TOOL_OUTPUT_REF_KEY] !== key &&
+    obj[TOOL_OUTPUT_REF_KEY] != null
+  ) {
+    return null;
+  }
+
+  const injected: Record<string, unknown> = {
+    [TOOL_OUTPUT_REF_KEY]: key,
+    ...obj,
+  };
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/tools/toolOutputReferences.ts]: Fixed in 8e0973c — reversed the merge to `{ ...obj, [TOOL_OUTPUT_REF_KEY]: key }` so the injected key lands after the spread and always wins. The guard that lets `_ref: null` pass through (treating it as non-conflicting) is preserved, but the null can no longer clobber our generated tag. Added a test case that registers a `{"_ref":null,...}` output and asserts the annotated form carries the correct `tool<i>turn<n>` value.

**Comment by danny-avila** [comment]
@codex review

File: src/tools/ToolNode.ts [hunk 3]
```diff
@@ -247,24 +310,58 @@ export class ToolNode<T = any> extends RunnableCallable<T, T> {
...
-            this.maxToolResultChars
-          ),
-          tool_call_id: call.id!,
-        });
       }
+      if (isBaseMessage(output) && output._getType() === 'tool') {
+        const toolMsg = output as ToolMessage;
+        if (
+          toolMsg.status !== 'error' &&
+          (this.toolOutputRegistry != null || unresolvedRefs.length > 0)
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/tools/ToolNode.ts]: Fixed in 342fc8b — `[unresolved refs: …]` now gets appended to error ToolMessages across all three error paths: thrown exceptions in `runTool`, tool-returned `ToolMessage({status:'error'})`, and host-returned error results in `dispatchToolEvents`. Registration/annotation of the reference key itself still stays off for errors (they shouldn't be pipeable), but the self-correction signal is now visible. Added a test for each of the three paths.

**Comment by danny-avila** [comment]
@codex review

File: src/tools/ToolNode.ts [hunk 4]
```diff
@@ -833,14 +1067,22 @@ export class ToolNode<T = any> extends RunnableCallable<T, T> {
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   protected async run(input: any, config: RunnableConfig): Promise<T> {
     this.toolCallTurns.clear();
+    const incomingRunId = config.configurable?.run_id as string | undefined;
+    if (incomingRunId !== this.lastRunId) {
+      this.turnCounter = 0;
+      this.toolOutputRegistry?.clear();
+      this.warnedNonStringRefTools.clear();
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/tools/ToolNode.ts]: Fixed in efa2a8e — when `config.configurable.run_id` is absent we now treat every batch as a fresh run and clear the registry + warn-set + turn counter on entry. Production callers (anyone going through `Run`) always set `run_id`, so this only changes behavior for direct ToolNode usage without one, which is exactly the cross-session-reuse case you flagged. Added a test proving that anonymous callers no longer see cross-batch `{{tool<i>turn<n>}}` resolution (they get `[unresolved refs: …]` instead). Updated the existing cross-batch tests to pass an explicit `run_id` through the test helper so accumulation behavior stays exercised under the new semantics.

**Comment by danny-avila** [comment]
@codex review

File: src/tools/ToolNode.ts [hunk 5]
```diff
@@ -309,15 +437,51 @@ export class ToolNode<T = any> extends RunnableCallable<T, T> {
...
+   */
+  private applyOutputReference(
+    truncated: string,
+    batchIndex: number | undefined,
+    unresolved: string[]
+  ): string {
+    let content = truncated;
+    if (this.toolOutputRegistry != null && batchIndex != null) {
+      const key = buildReferenceKey(batchIndex, this.currentTurn);
+      this.toolOutputRegistry.set(key, truncated);
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/tools/ToolNode.ts]: Fixed in 1a8414b. `run()` now captures `turn` in a local at the top of the sync prefix and threads it down through `runTool` / `dispatchToolEvents`; `applyOutputReference` takes a precomputed `refKey` string instead of reading `this.currentTurn` after any await, so concurrent `invoke()` calls cannot step on each other's key. Added a test that overlaps two batches of the same run (gates resolved in reverse order) and asserts each batch registers under its own `tool0turn0` / `tool0turn1` with correct content. The previous code would have baked the same turn value into both results. `this.currentTurn` remains as a read-only observability field; nothing on the registration path reads from it anymore.

Separately, the registry now preserves the raw untruncated output (5 MB total cap) so `{{…}}` substitutions can pipe full payloads into downstream tools even when the LLM saw a truncated preview — tracked as a parallel change in the same commit.

**Comment by danny-avila** [comment]
@codex review

File: src/tools/ToolNode.ts [hunk 6]
```diff
@@ -133,6 +168,34 @@ export class ToolNode<T = any> extends RunnableCallable<T, T> {
...
+       * `maxToolResultChars`: the registry stores the raw untruncated
+       * output so a later `{{…}}` substitution pipes the full payload
+       * into the next tool, even when the LLM saw a truncated preview.
+       * Pass the caller's config through unchanged — the registry fills
+       * in its own 5 MB defaults.
+       */
+      this.toolOutputRegistry = new ToolOutputReferenceRegistry({
+        maxOutputSize: toolOutputReferences.maxOutputSize,
+        maxTotalSize: toolOutputReferences.maxTotalSize,
+      });
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/tools/ToolNode.ts:228]: Fixed in 1b62fd4. The registry is now shared across every ToolNode a `Graph` compiles — `Graph` lazily constructs one instance per run via `getOrCreateToolOutputRegistry()` and threads it into each `new CustomToolNode(...)` through a new `toolOutputRegistry` option. I also moved the run-scoped state (turn counter, last-seen runId, warn-once memo) off of ToolNode and onto the registry itself via `nextTurn(runId)` / `claimWarnOnce(toolName)`, so cross-agent keys stay globally unique (Agent A's first tool is `tool0turn0`, Agent B's first tool is `tool0turn1`) and warn lines don't double-emit. Added a test that registers output in one ToolNode and resolves `{{tool0turn0}}` in a different ToolNode sharing the same registry — cross-node substitution works and the second node's output lands under the next shared turn. Direct ToolNode usage (no graph) still gets a local registry from the config object as before.

**Comment by danny-avila** [comment]
@codex review

File: src/tools/ToolNode.ts [hunk 7]
```diff
@@ -309,15 +450,62 @@ export class ToolNode<T = any> extends RunnableCallable<T, T> {
...
+    refKey: string | undefined,
+    unresolved: string[]
+  ): string {
+    let content = llmContent;
+    if (this.toolOutputRegistry != null && refKey != null) {
+      this.toolOutputRegistry.set(refKey, registryContent);
+      content = annotateToolOutputWithReference(llmContent, refKey);
+    }
+    if (unresolved.length > 0) {
+      content += `\n[unresolved refs: ${unresolved.join(', ')}]`;
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/tools/ToolNode.ts]: Fixed in 233cd29. `annotateToolOutputWithReference` now takes the `unresolved[]` list and routes both the ref key and the unresolved refs through the same JSON injection path when the output is a parseable object — unresolved keys land in a new `_unresolved_refs` array field alongside `_ref`, keeping the annotated content valid JSON for downstream consumers. Non-object content keeps the old `[ref: …]\n…\n[unresolved refs: …]` text form. Added tests covering JSON preservation under unresolved-with-ref, unresolved-only (error paths), and no-op-when-nothing-to-annotate cases.

File: src/tools/toolOutputReferences.ts [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+   * off-module so callers of the exported `TOOL_OUTPUT_REF_PATTERN`
+   * never see a stale `lastIndex`.
+   */
+  private static readonly PLACEHOLDER_MATCHER = /\{\{(tool\d+turn\d+)\}\}/g;
+
+  constructor(options: ToolOutputReferenceRegistryOptions = {}) {
+    const perOutput =
+      options.maxOutputSize != null && options.maxOutputSize > 0
+        ? options.maxOutputSize
+        : HARD_MAX_TOTAL_TOOL_OUTPUT_SIZE;
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/tools/toolOutputReferences.ts]: Fixed in 233cd29. Default per-output cap is now `HARD_MAX_TOOL_RESULT_CHARS` (~400 KB), matching the standard tool-result truncation budget and keeping `{{…}}` expansions within typical shell `ARG_MAX` limits. The 5 MB total cap is preserved so hosts that want fuller fidelity (non-bash API consumers, long JSON streams) can raise `maxOutputSize` explicitly via config — the JSDoc on `ToolOutputReferencesConfig.maxOutputSize` now calls out the shell-limit trade-off.

**Comment by danny-avila** [comment]
@codex review

File: src/tools/ToolNode.ts [hunk 9]
```diff
@@ -186,7 +262,12 @@ export class ToolNode<T = any> extends RunnableCallable<T, T> {
       if (call.id != null && call.id !== '') {
         this.toolCallTurns.set(call.id, turn);
       }
-      const args = call.args;
+      let args = call.args;
+      if (registry != null) {
+        const { resolved, unresolved } = registry.resolve(args);
+        args = resolved;
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/tools/ToolNode.ts:341]: Fixed in 0da124f. ToolNode now tracks `resolvedArgsByCallId` (cleared at the start of every batch, populated right after `registry.resolve` when substitution actually changed the args). `handleRunToolCompletions` prefers that map and falls back to `call.args` when no substitution happened, so `ON_RUN_STEP_COMPLETED.tool_call.args` carries what the tool received rather than the `{{…}}` template. The event-driven dispatch path already emitted resolved args via `entry.args` so no change was needed there. Added a test asserting the post-substitution command is what lands in the step-completed event.

File: src/tools/ToolNode.ts [hunk 10]
```diff
@@ -247,24 +328,84 @@ export class ToolNode<T = any> extends RunnableCallable<T, T> {
...
+              llmContent,
+              rawContent,
+              refKey,
+              unresolvedRefs
+            );
+          } else if (refKey != null) {
+            /**
+             * Known limitation: tools that return a `ToolMessage` with
+             * array-type content (multi-part content blocks such as
+             * text + image) are not registered under a reference key
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/tools/ToolNode.ts]: Fixed in 0da124f. When a tool returns `ToolMessage` with array-type content and the args had unresolved refs, we now prepend a `{ type: 'text', text: '[unresolved refs: …]' }` block to the content array so the LLM gets the same self-correction signal the string and error paths emit. Original content blocks are preserved after the warning. Registration of the reference key itself is still skipped (multi-part content has no canonical serialized form), with the existing warn-once log line. Added a test that exercises a multi-part-content tool and asserts the leading warning block.

**Comment by danny-avila** [comment]
@codex review

File: src/tools/toolOutputReferences.ts [hunk 11]
```diff
@@ -0,0 +1,427 @@
...
+   * reaches here is either null or matching) so we can place our
+   * injected keys *first* in the serialized JSON. Leading-key
+   * placement means the LLM sees the reference identifier before
+   * it reads the payload.
+   */
+  const {
+    [TOOL_OUTPUT_REF_KEY]: _existingRef,
+    [TOOL_OUTPUT_UNRESOLVED_KEY]: _existingUnresolved,
+    ...rest
+  } = obj;
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/tools/toolOutputReferences.ts]: Fixed in 113b4b8. `tryInjectRefIntoJsonObject` now only strips the framework-owned key it's actually injecting — a pre-existing `_unresolved_refs` payload field is preserved on the ref-only path, and a pre-existing `_ref` is preserved on the unresolved-only/error path. Also added a deep-equal collision guard for `_unresolved_refs` mirroring the existing `_ref` guard: a non-matching existing array falls back to the prefix form rather than silently overwriting, and a deep-equal array is accepted idempotently. Covered the four cases with new tests (ref-only preserves unresolved, unresolved-only preserves ref, collision fallback, idempotent match).

**Comment by danny-avila** [comment]
@codex review

File: src/tools/ToolNode.ts [hunk 12]
```diff
@@ -900,21 +1242,41 @@ export class ToolNode<T = any> extends RunnableCallable<T, T> {
...
+        const eventCalls = eventEntries.map((e) => e.call);
+        const eventIndices = eventEntries.map((e) => e.batchIndex);
 
         const directOutputs: (BaseMessage | Command)[] =
           directCalls.length > 0
             ? await Promise.all(
-              directCalls.map((call) => this.runTool(call, config))
+              directCalls.map((call, i) =>
+                this.runTool(call, config, directIndices[i], turn)
+              )
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/tools/ToolNode.ts:1457]: Fixed in 8c426b3. Direct and event branches now run in parallel via `Promise.all([directPromise, eventPromise])`. Both branches' sync prefixes — where all `registry.resolve(args)` passes live — complete against the *pre-batch* registry state before either branch's microtasks (and therefore registrations) get a turn. JS single-threaded execution guarantees no registration runs between the two sync prefixes, so same-turn refs stay unresolved regardless of how a batch splits across direct/event. Added a test that confirms a batch with a direct call at index 0 and an event call at index 1 referencing `{{tool0turn0}}` produces a host request with the literal template (no same-turn resolution), while the next batch resolves the ref normally across the boundary.

**Comment by danny-avila** [comment]
@codex review

File: src/tools/ToolNode.ts [hunk 13]
```diff
@@ -900,35 +1242,75 @@ export class ToolNode<T = any> extends RunnableCallable<T, T> {
...
+            ? this.dispatchToolEvents(eventCalls, config, eventIndices, turn)
+            : Promise.resolve({
               toolMessages: [] as ToolMessage[],
               injected: [] as BaseMessage[],
-            };
+            });
+
+        const [directOutputs, eventResult] = await Promise.all([
+          directPromise,
+          eventPromise,
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/tools/ToolNode.ts]: Fixed in 4926e07. I snapshot event args synchronously against the pre-batch registry state in `run()`, then serialize: directs await first, events dispatch only if directs succeeded. Same-turn isolation still holds because event args were captured pre-await, but a thrown direct tool (handleToolErrors=false, GraphInterrupt) now aborts before any host dispatch happens. Added a test with `handleToolErrors: false` + a throwing direct tool + a mock `safeDispatchCustomEvent` spy — assert the rejection and confirm the host was never called.

File: src/tools/toolOutputReferences.ts [hunk 14]
```diff
@@ -0,0 +1,461 @@
...
+   * the registry, warn-set, and counter are cleared before claiming
+   * turn 0. Missing `runId` is treated as "always a new run" so
+   * anonymous callers never leak state across invocations.
+   */
+  nextTurn(runId: string | undefined): number {
+    if (runId == null || runId !== this.lastRunId) {
+      this.entries.clear();
+      this.totalSize = 0;
+      this.turnCounter = 0;
+      this.warnedNonStringTools.clear();
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/tools/toolOutputReferences.ts]: Fixed in 4926e07. The registry is now internally partitioned by `runId`: each bucket owns its own entries/totalSize/turnCounter/warnMemo. Every public method (`set`/`get`/`resolve`/`nextTurn`/`claimWarnOnce`) takes a `runId` and routes to the correct bucket; ToolNode threads runId through every registry call site. Anonymous (`run_id` absent) batches still reset per-call via a sentinel bucket that gets dropped on entry. Added `releaseRun(runId)` for deterministic cleanup and a `maxActiveRuns` option (default 32) with FIFO eviction of oldest bucket to keep memory bounded when many short-lived runs accumulate. Added an isolation test: two runs back-to-back on a shared registry, verifying run A's output survives run B starting and each run resolves `{{tool0turn0}}` against its own partition.

**Comment by danny-avila** [comment]
@codex review

File: src/graphs/Graph.ts [hunk 15]
```diff
@@ -153,8 +167,31 @@ export abstract class Graph<
     this.invokedToolIds = undefined;
     this.handlerRegistry = undefined;
     this.hookRegistry = undefined;
+    this.toolOutputReferences = undefined;
+    this._toolOutputRegistry = undefined;
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/graphs/Graph.ts:180]: Fixed in 475403d. `Graph.clearHeavyState` now calls `this._toolOutputRegistry?.clear()` before nulling its own reference, so stored `tool<i>turn<n>` entries (plus the per-run turn counters and warn memos) are wiped — any ToolNode still holding the captured registry reference sees an empty registry on the next `processStream()` call. Added unit tests for both `clear()` (drops every bucket) and `releaseRun(runId)` (drops only the named run, leaves others intact).

File: src/tools/toolOutputReferences.ts [hunk 16]
```diff
@@ -0,0 +1,519 @@
...
+   * distinct turn values (reads are effectively atomic by JS's
+   * single-threaded execution of the sync prefix).
+   *
+   * If `runId` is missing the anonymous bucket is dropped and a
+   * fresh one created so each anonymous call behaves as its own run.
+   */
+  nextTurn(runId: string | undefined): number {
+    if (runId == null) {
+      this.runStates.delete(ANON_RUN_KEY);
+    }
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/tools/toolOutputReferences.ts:294]: Fixed in 475403d. ToolNode now mints a unique `batchScopeId` (of the form `\0anon-<N>` via a monotonic counter) for every anonymous batch and threads it through `runTool`/`dispatchToolEvents`/`applyOutputReference` in place of the raw `run_id`. Each anonymous invocation gets its own registry partition so two concurrent `node.invoke(...)` calls without `run_id` can't clobber each other's in-flight state. Added a concurrent-anon test: two parallel invokes, gates resolved in reverse order — each observes its own `[ref: tool0turn0]` with no cross-contamination.

**Comment by danny-avila** [comment]
@codex review

File: src/tools/toolOutputReferences.ts [hunk 17]
```diff
@@ -0,0 +1,519 @@
...
+     */
+    const perOutput =
+      options.maxOutputSize != null && options.maxOutputSize > 0
+        ? options.maxOutputSize
+        : HARD_MAX_TOOL_RESULT_CHARS;
+    const totalRaw =
+      options.maxTotalSize != null && options.maxTotalSize > 0
+        ? options.maxTotalSize
+        : calculateMaxTotalToolOutputSize(perOutput);
+    this.maxTotalSize = totalRaw;
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/tools/toolOutputReferences.ts:188]: Fixed in 50904fe. The user-provided branch now applies `Math.min(options.maxTotalSize, HARD_MAX_TOTAL_TOOL_OUTPUT_SIZE)` so the documented 5 MB cap holds regardless of what hosts pass. Per-output stays clamped to the (now hard-bounded) total via the existing `Math.min(perOutput, totalRaw)`, so a request like `{ maxTotalSize: 50_000_000 }` ends up with both `totalLimit` and `perOutputLimit` ≤ 5 MB. Added a test that exercises exactly that case.

**Comment by danny-avila** [comment]
@codex review

File: src/tools/toolOutputReferences.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/tools/toolOutputReferences.ts:188]: Fixed in dd318f2. Constructor now wraps the user-provided branch in `Math.min(options.maxTotalSize, HARD_MAX_TOTAL_TOOL_OUTPUT_SIZE)` so the documented 5 MB absolute cap is enforced regardless of caller config — the computed-default branch already applied the same upper bound via `calculateMaxTotalToolOutputSize`. Added a unit test asserting that a 50 MB request gets clamped down to 5 MB on both `totalLimit` and `perOutputLimit`.

File: src/tools/ToolNode.ts [hunk 18]
```diff
@@ -565,7 +875,30 @@ export class ToolNode<T = any> extends RunnableCallable<T, T> {
...
+           * copying user-supplied text) that the pre-hook pass never
+           * saw. Re-running the resolver on the hook-rewritten args
+           * keeps substitution and the unresolved-refs record in sync
+           * with what the tool will actually receive.
+           */
+          if (registry != null) {
+            const { resolved, unresolved } = registry.resolve(
+              registryRunId,
+              hookResult.updatedInput
+            );
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/tools/ToolNode.ts:953]: Fixed in dd318f2. Added `ToolOutputReferenceRegistry.snapshot(runId)` returning a `ToolOutputResolveView` that resolves placeholders against a defensive copy of the bucket's entries. The mixed direct+event path now takes a snapshot synchronously before launching directs and threads it into `dispatchToolEvents` (new `preBatchSnapshot` parameter). The PreToolUse hook re-resolution uses the snapshot when present — so a hook-introduced `{{tool<i>turn<current>}}` resolves only against pre-batch state and cannot pick up same-turn direct outputs that registered while the event branch was awaiting. Pure-event batches still resolve against the live registry (correct because no in-batch registrations have happened yet by hook time). Added an integration test: mixed batch with a `PreToolUse` hook that injects `{{tool0turn0}}` — host receives the literal template, confirming the same-turn direct output is invisible.

**Comment by danny-avila** [comment]
@codex review

**Comment by danny-avila** [comment]
Audit follow-ups for the 4 findings from the comprehensive review pushed in aa0952b:

- **#1 (param explosion)**: Replaced positional optional params with two typed options bags — `RunToolBatchContext` and `DispatchBatchContext`. `runTool` is now `(call, config, batchContext)` (3 params), `dispatchToolEvents` and `executeViaEvent` are `(toolCalls, config, [input,] batchContext)`. Eliminates the `undefined`-placeholder anti-pattern in `executeViaEvent` and stops future fields from re-bloating the signatures.
- **#2 (import-type grouping)**: Moved `import type { ToolOutputResolveView, … }` up into the existing type-import block per AGENTS.md ordering rules. Also exported `PreResolvedArgsMap` and `ResolvedArgsByCallId` from the registry so the new context types reuse one source of truth.
- **#3 (test colocation)**: Moved `src/utils/truncation.test.ts` → `src/utils/__tests__/truncation.test.ts`, switched the import to the `@/utils/truncation` alias to match the project's `__tests__/` convention.
- **#4 (missing snapshot edge case)**: Added a unit test for `snapshot('never-touched')` confirming the empty-bucket fallback returns the original args and reports placeholders as unresolved.

Also addressed the two new round-14 P2 codex inline findings in the previous push (dd318f2): caller-supplied `maxTotalSize` clamped to `HARD_MAX_TOTAL_TOOL_OUTPUT_SIZE`, and a frozen `snapshot()` carrier prevents `PreToolUse`-injected placeholders from leaking same-turn direct outputs in the mixed direct+event path.

469 tests passing. `tsc --noEmit` clean. `eslint` clean on changed files.

@codex review

**Comment by danny-avila** [comment]
Pushed [443fcdc](https://github.com/danny-avila/agents/commit/443fcdc): applied the one NIT (typed `handleRunToolCompletions`'s `resolvedArgsByCallId` parameter via the exported `ResolvedArgsByCallId` alias, moved its docstring to a `@param` block) and bumped `@librechat/agents` to `3.1.70-dev.0` for the tool-output-references release. 469 tests pass, typecheck/lint clean.
